diff -up --recursive --new-file mopd-2.5.3.macro/common/pf.c mopd-2.5.3/common/pf.c
--- mopd-2.5.3.macro/common/pf.c	1996-08-06 14:20:27.000000000 +0000
+++ mopd-2.5.3/common/pf.c	2002-11-17 15:22:31.000000000 +0000
@@ -65,7 +65,7 @@ static char rcsid[] = "$Id: pf.c,v 1.16 
  */
 
 extern int errno;
-extern int promisc;
+extern int nomulti;
 
 /*
  * Return information to device.c how to open device.
@@ -144,7 +144,8 @@ pfInit(interface, mode, protocol, typ)
       		syslog(LOG_ERR,"pfInit: %s is not ethernet", device);
 		return(-1);
 	}
-	if (promisc) {
+	if (!nomulti) {
+		syslog(LOG_WARNING,"pfInit: using promiscuous mode for multicast reception");
 		/* Set promiscuous mode. */
 		if (ioctl(fd, BIOCPROMISC, (caddr_t)0) < 0) {
       			syslog(LOG_ERR,"pfInit: BIOCPROMISC: %m");
