diff -up --recursive --new-file mopd-2.5.3.macro/common/device.c mopd-2.5.3/common/device.c
--- mopd-2.5.3.macro/common/device.c	2001-10-27 19:14:58.000000000 +0000
+++ mopd-2.5.3/common/device.c	2002-11-17 12:50:33.000000000 +0000
@@ -151,12 +151,7 @@ deviceOpen(ifname, proto, trans)
 #ifdef	DEV_NEW_CONF
 		deviceEthAddr(p->if_name,&p->eaddr[0]);
 #elif	defined(__linux__)
-		{
-			int s;
-			s = socket(AF_INET,SOCK_DGRAM,0);
-			pfEthAddr(s,p->if_name,&p->eaddr[0]);
-			(void) close(s);
-		}
+		pfEthAddr(0, p->if_name,&p->eaddr[0]);
 #else
 		p->eaddr[0]= tmp.eaddr[0];
 		p->eaddr[1]= tmp.eaddr[1];
diff -up --recursive --new-file mopd-2.5.3.macro/common/pf-linux.c mopd-2.5.3/common/pf-linux.c
--- mopd-2.5.3.macro/common/pf-linux.c	2002-11-17 12:38:31.000000000 +0000
+++ mopd-2.5.3/common/pf-linux.c	2002-11-17 12:50:38.000000000 +0000
@@ -178,12 +178,17 @@ int s;
 char *interface;
 u_char *addr;
 {
+  int err;
+
   strncpy(ifr.ifr_name, interface, sizeof (ifr.ifr_name) - 1);
   ifr.ifr_name[sizeof(ifr.ifr_name)] = 0;
   ifr.ifr_addr.sa_family = AF_INET;
-  if (ioctl(s, SIOCGIFHWADDR, &ifr) < 0) {
+  s = socket(AF_INET,SOCK_DGRAM,0);
+  err = ioctl(s, SIOCGIFHWADDR, &ifr);
+  (void) close(s);
+  if (err < 0) {
     syslog(LOG_ERR, "pfEthAddr: %s: SIOCGIFHWADDR: %m", interface);
-    return(-1);
+    exit(-1);
   }
   memcpy((char *)addr, ifr.ifr_hwaddr.sa_data, 6);
   return(0);
